### ISSUE TITLE
TSLint asks to annotate the props variable in the constructor of an annotated React class

### ISSUE THREAD

**Issue by wildeyes** [palantir/tslint#2674]
![image](https://cloud.githubusercontent.com/assets/1652108/25661456/64a2bc50-3019-11e7-8104-e73a49fd8190.png)

### Bug Report

- __TSLint version__: 4.3.1
- __TypeScript version__: 2.3.2
- __Running TSLint via__:  Visual Studio Code

#### TypeScript code being linted

```ts
interface Props {}
class X extends Component<Props, void> {
  constructor(props) {
}
render() { return null; }
}
```

#### Actual behavior
Asks me to annotate props.

#### Expected behavior
figure out by itself that props is equal to the props of the react component.

**Comment by andy-hanson** [comment]
Try turning on [`--noImplicitAny`](https://www.typescriptlang.org/docs/handbook/compiler-options.html). TypeScript can't figure out the type of props, so neither can tslint!

**Comment by adidahiya** [comment]
Well, `typedef` doesn't use the type checker -- it only does syntactic checks. Currently there aren't a lot of "smart" options for that rule, and adding them is a pretty low priority for us. See discussions like https://github.com/palantir/tslint/issues/669

**Comment by JoshuaKGoldberg** [comment]
☠️ *TSLint's time has come!* ☠️ 

TSLint is no longer accepting most feature requests per #4534. See **[typescript-eslint.io](https://typescript-eslint.io)** for the new, shiny way to lint your TypeScript code with **ESLint**. ✨ 

It was a pleasure open sourcing with you all!
